core/src/connection.rs

```diff
@@ -0,0 +1,400 @@
+    /// since the connection has been established.
+    ///
+    /// See also [`StreamMuxer::is_remote_acknowledged`].
+    pub fn is_remote_acknowledged(&self) -> bool {
+        self.muxing.is_remote_acknowledged()
+    }
+
+    /// Begins an orderly shutdown of the connection, returning a
+    /// `Future` that resolves when connection shutdown is complete.
+    pub fn close(self) -> Close<TMuxer> {
+        self.muxing.close().0
+    }
+
+    /// Polls the connection for events produced by the associated handler
+    /// as a result of I/O activity on the substream multiplexer.
+    pub fn poll<TTransErr>(mut self: Pin<&mut Self>, cx: &mut Context)
+        -> Poll<Result<THandler::OutEvent, ConnectionError<THandler::Error, TTransErr>>>
+    {
+        loop {
+            let mut io_pending = false;
+
+            // Perform I/O on the connection through the muxer, informing the handler
+            // of new substreams.
+            match self.muxing.poll(cx) {
+                Poll::Pending => io_pending = true,
+                Poll::Ready(Ok(SubstreamEvent::InboundSubstream { substream })) => {
+                    self.handler.inject_substream(substream, SubstreamEndpoint::Listener)
+                }
+                Poll::Ready(Ok(SubstreamEvent::OutboundSubstream { user_data, substream })) => {
+                    let endpoint = SubstreamEndpoint::Dialer(user_data);
+                    self.handler.inject_substream(substream, endpoint)
+                }
+                Poll::Ready(Err(err)) => return Poll::Ready(Err(ConnectionError::IO(err))),
+            }
+
+            // Poll the handler for new events.
+            match self.handler.poll(cx) {
+                Poll::Pending => {
+                    if io_pending {
+                        return Poll::Pending // Nothing to do
+                    }
+                }
+                Poll::Ready(Ok(ConnectionHandlerEvent::OutboundSubstreamRequest(user_data))) => {
+                    self.muxing.open_substream(user_data);
+                }
+                Poll::Ready(Ok(ConnectionHandlerEvent::Custom(event))) => {
+                    return Poll::Ready(Ok(event));
+                }
+                Poll::Ready(Err(err)) => return Poll::Ready(Err(ConnectionError::Handler(err))),
+            }
+        }
+    }
+}
+
+/// Attempt to reach a peer.
+#[derive(Debug, Clone)]
+pub struct DialingAttempt {
+    /// Identifier for the reach attempt.
+    pub(crate) id: ConnectionId,
+    /// Multiaddr currently being attempted.
+    pub(crate) current: Multiaddr,
+    /// Multiaddresses to attempt if the current one fails.
+    pub(crate) next: Vec<Multiaddr>,
+}
+
+/// A `DialingConnection` is a [`PendingConnection`] where the local peer
+/// has the role of the dialer (i.e. initiator) and the (expected) remote
+/// peer ID is known.
+pub struct DialingConnection<'a, TInEvent, TConnInfo, TPeerId> {
+    pub(crate) peer_id: &'a TPeerId,
```

Comment:
The `DialingConnection` is only constructed in the `network::peer` module. I was going back and forth between putting in in the `connection` module or `network::peer` module. I could either create a `pub(crate)` constructor here instead (since the `connection` module is public) or indeed move the whole definition over to the `network::peer` module. I suppose you prefer the latter?

Reply:
I would indeed move the definition, if that's ok. At the very least, use `pub(super)` rather than `pub(crate)`.

Reply:
Done.